File: 1386-cinema-seat-allocation/1386-cinema-seat-allocation.go
func maxNumberOfFamilies(n int, reservedSeats [][]int) int {
    ans := 2 * n // max possible result if all seats aren't reserved

    sort.Slice(reservedSeats, func(i, j int) bool {
        return reservedSeats[i][0] < reservedSeats[j][0]
    })

    idx := 0
    for idx < len(reservedSeats) {
        currRow := reservedSeats[idx][0]
        // left: s2, s3 reserved
        // midLeft: s4, s5 reserved
        // midRight: s6, s7 reserved
        // right: s8, s9 reserved
        left, midLeft, midRight, right := false, false, false, false
        for idx < len(reservedSeats) && reservedSeats[idx][0] == currRow {
            seat := reservedSeats[idx][1]
            if seat == 2 || seat == 3 {
                left = true
            } else if seat == 4 || seat == 5 {
                midLeft = true
            } else if seat == 6 || seat == 7 {
                midRight = true
            } else if seat == 8 || seat == 9 {
                right = true
            }

            idx++
        }

        if (left && midRight) || (midLeft && midRight) || (midLeft && right) {
            ans -= 2
        } else if left || midLeft || midRight || right {
            ans -= 1
        }
    }

    return ans
}

// array, greedy
// time: O(nlog(n))
// space: O(1)

File: 1386-cinema-seat-allocation/1386-cinema-seat-allocation.py
class Solution:
    def maxNumberOfFamilies(self, n: int, reservedSeats: List[List[int]]) -> int:
        ans = 2 * n

        reservedSeats.sort()

        idx = 0
        while idx < len(reservedSeats):
            curr_row = reservedSeats[idx][0]
            left, mid_left, mid_right, right = False, False, False, False
            while idx < len(reservedSeats) and reservedSeats[idx][0] == curr_row:
                seat = reservedSeats[idx][1]

                if seat == 2 or seat == 3:
                    left = True
                elif seat == 4 or seat == 5:
                    mid_left = True
                elif seat == 6 or seat == 7:
                    mid_right = True
                elif seat == 8 or seat == 9:
                    right = True
                
                idx += 1
            
            if (left and mid_right) or (mid_left and mid_right) or (mid_left and right):
                ans -= 2
            elif left or mid_left or mid_right or right:
                ans -= 1
        
        return ans

# array, greedy, sorting
# time: O(nlog(n))
# space: O(1)